File: packages/backend-kotlin/native/LucentRuntime.kt
{{error}}

{{errorWire}}

object LucentBytes {
  fun toByteArray(buffer: ArrayBuffer): ByteArray {
    {{bytesToByteArray}}
  }
  fun fromByteArray(bytes: ByteArray): ArrayBuffer {
    {{bytesFromByteArray}}
  }
  fun length(buffer: ArrayBuffer): Double {
    {{bytesLength}}
  }

  fun get(buffer: ArrayBuffer, index: Double): Double {
    {{bytesGet}}
  }
}

fun lucentStr(value: Double): String {
  if (value.isFinite() && value == Math.rint(value) && Math.abs(value) < 1e15) {
    return value.toLong().toString()
  }
  return value.toString()
}

fun lucentStr(value: Float): String = lucentStr(value.toDouble())
fun lucentStr(value: Int): String = value.toString()
fun lucentStr(value: Long): String = value.toString()
fun lucentStr(value: Short): String = value.toString()
fun lucentStr(value: Byte): String = value.toString()
fun lucentStr(value: UInt): String = value.toString()
fun lucentStr(value: ULong): String = value.toString()
fun lucentStr(value: UShort): String = value.toString()
fun lucentStr(value: UByte): String = value.toString()
fun lucentStr(value: Boolean): String = if (value) "true" else "false"

File: packages/backend-swift/native/LucentRuntime.swift
import Foundation

{{error}}

{{errorWire}}

enum LucentBytes {
  static func data(_ buffer: ArrayBuffer) -> Data {
    {{bytesData}}
  }
  static func fromData(_ data: Data) throws -> ArrayBuffer {
    {{bytesFromData}}
  }
  static func length(_ buffer: ArrayBuffer) -> Double {
    {{bytesLength}}
  }

  static func get(_ buffer: ArrayBuffer, _ index: Double) -> Double {
    {{bytesGet}}
  }
}

func lucentStr(_ value: Double) -> String {
  if value.isFinite && value == value.rounded() && abs(value) < 1e15 {
    return String(Int64(value))
  }
  return String(value)
}

func lucentStr(_ value: Float) -> String {
  return lucentStr(Double(value))
}

func lucentStr<T: BinaryInteger>(_ value: T) -> String {
  return String(value)
}

func lucentStr(_ value: Bool) -> String {
  return value ? "true" : "false"
}
